In [1]:
import pandas as pd
import numpy as np

from mlxtend.frequent_patterns import apriori as apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import fpgrowth

## Load the Dataset

In [2]:
basket = pd.read_csv("basket.csv")
basket.columns

Index(['9 de oro', 'agua', 'agua tonica', 'alfajor block',
       'alfajor de maicena', 'alfajor santafesino', 'alka', 'aquarius',
       'baggio', 'bandeja de miga', 'barrita de cereales', 'belden',
       'bizcochos', 'block', 'brahma', 'café', 'campeon', 'caramelos',
       'cepita', 'cereales', 'chester', 'chupetin', 'coca', 'cofler',
       'conitos', 'corona', 'doble cola', 'doctor lemon', 'don satur',
       'doritos', 'dos corazones', 'encendedor', 'facturas', 'fanta',
       'fantoche', 'fernet', 'fresh', 'gancia', 'gatorade', 'gin', 'gomitas',
       'guaymayen', 'heineken', 'iguana', 'jorgito', 'kit kat', 'lata coca',
       'lays', 'levite', 'lincon', 'liverpool', 'lucky', 'malboro',
       'matecocido', 'mentoplus', 'mini oreo', 'mogul', 'monster', 'opera',
       'oreo', 'pan', 'paseo', 'pañuelos', 'pebete', 'petaca', 'phillip',
       'pier', 'pipas', 'pitusas', 'polvorita', 'postre', 'powerade', 'pritty',
       'pronto', 'quilmes', 'rhodesia', 'saladix', 'sanguche miga

## Brief Dataset Analysis

In [3]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1032 entries, 0 to 1031
Data columns (total 93 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   9 de oro             1032 non-null   bool 
 1   agua                 1032 non-null   bool 
 2   agua tonica          1032 non-null   bool 
 3   alfajor block        1032 non-null   bool 
 4   alfajor de maicena   1032 non-null   bool 
 5   alfajor santafesino  1032 non-null   bool 
 6   alka                 1032 non-null   bool 
 7   aquarius             1032 non-null   bool 
 8   baggio               1032 non-null   bool 
 9   bandeja de miga      1032 non-null   bool 
 10  barrita de cereales  1032 non-null   bool 
 11  belden               1032 non-null   bool 
 12  bizcochos            1032 non-null   bool 
 13  block                1032 non-null   bool 
 14  brahma               1032 non-null   bool 
 15  café                 1032 non-null   bool 
 16  campeon              103

## Association Rule Mining


"I will be using the association rule mining algorithms `apriori` and `fpgrowth` from the `mlxtend` library. These are the best algorithms for association rule mining.

Important metrics to remember:
- **Support**: It is the ratio of times the set/rule appears in the dataset. It is calculated as:

  $SP = \frac{A \cup B}{U}$

  Where $A$ is the antecedent of $B$, and they are items within the universe of the dataset $U$.

- **Confidence**: It is the ratio of times the consequent appears when the antecedent appears. It is calculated as:

  $CON = \frac{A \cup B}{A}$

- **Lift**: It compares the observed frequency of a rule with the expected frequency by chance (if the rule doesn't actually exist). It is calculated as:

  $LF = \frac{SP(A \cup B)}{SP(A) \times SP(B)}$"


### Calculating the Most Frequent Itemsets


In [4]:
mins = 0.001
freq_items_apri = apriori(basket,min_support=mins,use_colnames=True)
freq_items_fp = fpgrowth(basket,min_support=mins, use_colnames=True)

In [5]:
freq_items_apri.shape

(142, 2)

In [6]:
freq_items_fp.shape

(142, 2)

Both `apriori` and `fpgrowth` found 142 frequent itemsets with a support of 0.1%.


### Association Rule

#### Apriori

In [7]:
rules_apri = association_rules(freq_items_apri,metric="confidence",min_threshold=0.5)
rules_apri["antecedent_len"] = rules_apri["antecedents"].apply(lambda x: len(x))
rules_apri.sort_values("lift",ascending=False)

antecedents  consequents  antecedent support  \
3                  (matecocido)   (facturas)            0.005814   
0                    (pañuelos)     (belden)            0.002907   
5  (phillip, sanguche milanesa)       (coca)            0.001938   
6     (sanguche milanesa, coca)    (phillip)            0.002907   
1                     (conitos)  (caramelos)            0.002907   
2                        (lays)       (coca)            0.007752   
4          (phillip, guaymayen)       (coca)            0.003876   

   consequent support   support  confidence       lift  leverage  conviction  \
3            0.017442  0.002907    0.500000  28.666667  0.002806    1.965116   
0            0.055233  0.001938    0.666667  12.070175  0.001777    2.834302   
5            0.095930  0.001938    1.000000  10.424242  0.001752         inf   
6            0.074612  0.001938    0.666667   8.935065  0.001721    2.776163   
1            0.076550  0.001938    0.666667   8.708861  0.001715    2.770349   
2            0.095930  0.003876    0.500000   5.212121  0.003132    1.808140   
4            0.095930  0.001938    0.500000   5.212121  0.001566    1.808140   

   zhangs_metric  antecedent_len  
3       0.970760               1  
0       0.919825               1  
5       0.905825               2  
6       0.890671               2  
1       0.887755               1  
2       0.814453               1  
4       0.811284               2

#### Fpgrowth

In [8]:
rules_fp = association_rules(freq_items_fp,metric="confidence",min_threshold=0.7)
rules_fp["antecedent_len"] = rules_fp["antecedents"].apply(lambda x: len(x))
rules_fp.sort_values("lift",ascending=False)

antecedents consequents  antecedent support  \
0  (phillip, sanguche milanesa)      (coca)            0.001938   

   consequent support   support  confidence       lift  leverage  conviction  \
0             0.09593  0.001938         1.0  10.424242  0.001752         inf   

   zhangs_metric  antecedent_len  
0       0.905825               2

### Conclusion

Both algorithms found rules that are not very useful due to the significant noise in the data caused by those items that were sold infrequently. Additionally, the items "caramelos" and "alka" are not usually bought, they are given as change in a purchase. Therefore, I have decided to perform association rule mining again after removing the noise from my data.


## Association Rule Mining with cleaned dataset

In [9]:
basket_clean = pd.read_csv("basket_clean.csv")
basket_clean.shape

(794, 22)

As you can see, after removing the noise from my transactions, the data quantity reduced from 1034 to 794 transactions.


### Calculating the Most Frequent Itemsets

In [10]:
mins = 0.001
freq_items_apri = apriori(basket_clean,min_support=mins,use_colnames=True,max_len=3)
freq_items_fp = fpgrowth(basket_clean,min_support=mins, use_colnames=True,max_len=3)

In [11]:
freq_items_apri.shape

(121, 2)

In [12]:
freq_items_fp.shape

(121, 2)

Both `apriori` and `fpgrowth` found 121 frequent itemsets with a support of 0.1%.

### Association Rule

#### Apriori

In [13]:
rules_apri = association_rules(freq_items_apri,metric="confidence",min_threshold=0.5)
rules_apri["antecedent_len"] = rules_apri["antecedents"].apply(lambda x: len(x))
rules_apri.sort_values("lift",ascending=False)

antecedents          consequents  antecedent support  \
17                  (pier, levite)          (guaymayen)            0.001259   
22               (quilmes, levite)  (sanguche milanesa)            0.001259   
20    (sanguche milanesa, quilmes)             (levite)            0.002519   
18    (phillip, sanguche milanesa)          (guaymayen)            0.002519   
15               (guaymayen, pier)             (levite)            0.002519   
21     (sanguche milanesa, levite)            (quilmes)            0.001259   
3              (encendedor, lucky)             (belden)            0.001259   
5            (encendedor, malboro)             (belden)            0.001259   
6                 (lucky, malboro)             (belden)            0.001259   
4             (encendedor, belden)            (malboro)            0.002519   
19  (sanguche milanesa, guaymayen)            (phillip)            0.001259   
16             (guaymayen, levite)               (pier)            0.001259   
13    (phillip, sanguche milanesa)               (coca)            0.002519   
0                 (baggio, brahma)               (coca)            0.001259   
12             (liverpool, pebete)               (coca)            0.001259   
8             (chester, guaymayen)               (coca)            0.001259   
7              (chester, fantoche)               (coca)            0.001259   
11  (sanguche milanesa, guaymayen)               (coca)            0.001259   
2             (encendedor, belden)              (lucky)            0.002519   
1             (chester, mentoplus)             (belden)            0.002519   
14       (sanguche milanesa, coca)            (phillip)            0.003778   
10               (guaymayen, pier)               (coca)            0.002519   
9             (phillip, guaymayen)               (coca)            0.005038   

    consequent support   support  confidence       lift  leverage  conviction  \
17            0.027708  0.001259    1.000000  36.090909  0.001225         inf   
22            0.032746  0.001259    1.000000  30.538462  0.001218         inf   
20            0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
18            0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
15            0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
21            0.060453  0.001259    1.000000  16.541667  0.001183         inf   
3             0.071788  0.001259    1.000000  13.929825  0.001169         inf   
5             0.071788  0.001259    1.000000  13.929825  0.001169         inf   
6             0.071788  0.001259    1.000000  13.929825  0.001169         inf   
4             0.039043  0.001259    0.500000  12.806452  0.001161    1.921914   
19            0.096977  0.001259    1.000000  10.311688  0.001137         inf   
16            0.123426  0.001259    1.000000   8.102041  0.001104         inf   
13            0.124685  0.002519    1.000000   8.020202  0.002205         inf   
0             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
12            0.124685  0.001259    1.000000   8.020202  0.001102         inf   
8             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
7             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
11            0.124685  0.001259    1.000000   8.020202  0.001102         inf   
2             0.065491  0.001259    0.500000   7.634615  0.001094    1.869018   
1             0.071788  0.001259    0.500000   6.964912  0.001079    1.856423   
14            0.096977  0.002519    0.666667   6.874459  0.002152    2.709068   
10            0.124685  0.001259    0.500000   4.010101  0.000945    1.750630   
9             0.124685  0.002519    0.500000   4.010101  0.001891    1.750630   

    zhangs_metric  antecedent_len  
17       0.973518               2  
22       0.968474               2  
20       0.946970               2  
18       0.946970               2  
1

#### Fpgrowth

In [14]:
rules_fp = association_rules(freq_items_fp,metric="confidence",min_threshold=0.5)
rules_fp["antecedent_len"] = rules_fp["antecedents"].apply(lambda x: len(x))
rules_fp.sort_values("lift",ascending=False)

antecedents          consequents  antecedent support  \
19                  (pier, levite)          (guaymayen)            0.001259   
16               (quilmes, levite)  (sanguche milanesa)            0.001259   
14    (sanguche milanesa, quilmes)             (levite)            0.002519   
3     (phillip, sanguche milanesa)          (guaymayen)            0.002519   
17               (guaymayen, pier)             (levite)            0.002519   
15     (sanguche milanesa, levite)            (quilmes)            0.001259   
10           (encendedor, malboro)             (belden)            0.001259   
12             (encendedor, lucky)             (belden)            0.001259   
13                (lucky, malboro)             (belden)            0.001259   
9             (encendedor, belden)            (malboro)            0.002519   
4   (sanguche milanesa, guaymayen)            (phillip)            0.001259   
18             (guaymayen, levite)               (pier)            0.001259   
0              (chester, fantoche)               (coca)            0.001259   
20    (phillip, sanguche milanesa)               (coca)            0.002519   
22                (baggio, brahma)               (coca)            0.001259   
8              (liverpool, pebete)               (coca)            0.001259   
6             (chester, guaymayen)               (coca)            0.001259   
5   (sanguche milanesa, guaymayen)               (coca)            0.001259   
11            (encendedor, belden)              (lucky)            0.002519   
1             (chester, mentoplus)             (belden)            0.002519   
21       (sanguche milanesa, coca)            (phillip)            0.003778   
7                (guaymayen, pier)               (coca)            0.002519   
2             (phillip, guaymayen)               (coca)            0.005038   

    consequent support   support  confidence       lift  leverage  conviction  \
19            0.027708  0.001259    1.000000  36.090909  0.001225         inf   
16            0.032746  0.001259    1.000000  30.538462  0.001218         inf   
14            0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
3             0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
17            0.027708  0.001259    0.500000  18.045455  0.001190    1.944584   
15            0.060453  0.001259    1.000000  16.541667  0.001183         inf   
10            0.071788  0.001259    1.000000  13.929825  0.001169         inf   
12            0.071788  0.001259    1.000000  13.929825  0.001169         inf   
13            0.071788  0.001259    1.000000  13.929825  0.001169         inf   
9             0.039043  0.001259    0.500000  12.806452  0.001161    1.921914   
4             0.096977  0.001259    1.000000  10.311688  0.001137         inf   
18            0.123426  0.001259    1.000000   8.102041  0.001104         inf   
0             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
20            0.124685  0.002519    1.000000   8.020202  0.002205         inf   
22            0.124685  0.001259    1.000000   8.020202  0.001102         inf   
8             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
6             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
5             0.124685  0.001259    1.000000   8.020202  0.001102         inf   
11            0.065491  0.001259    0.500000   7.634615  0.001094    1.869018   
1             0.071788  0.001259    0.500000   6.964912  0.001079    1.856423   
21            0.096977  0.002519    0.666667   6.874459  0.002152    2.709068   
7             0.124685  0.001259    0.500000   4.010101  0.000945    1.750630   
2             0.124685  0.002519    0.500000   4.010101  0.001891    1.750630   

    zhangs_metric  antecedent_len  
19       0.973518               2  
16       0.968474               2  
14       0.946970               2  
3        0.946970               2  
1

In this case, fewer rules were obtained, but in my opinion, these have a lower level of noise compared to the ones obtained previously.


## Interesting Rules
